### PR TITLE
[mini] Improve native crash handling.

### PR DESCRIPTION
**Pull request by alexrp**
The most important change here is that we now do diagnostics in a different order: `/proc/self/maps`, GDB, `backtrace ()`, then `mono_walk_stack ()`.

If the process is in a really bad shape after a native crash, there's a good chance that `mono_walk_stack ()` would just crash, which would trigger the crash loop detection and abandon all attempts at further diagnostics, including GDB.

Now, we do GDB and `backtrace ()` before we call `mono_walk_stack ()` so that even if things are broken to the point where we can't get a managed backtrace, at least we can get the (usually much more useful) GDB info.

Additionally, if we get a crash while diagnosing a crash, we try to print a hint about which step failed. There's also better handling of error return values from some system calls.

### CODE REVIEW

**Comment by vargaz** [comment]
gdb/lldb is also not 100% reliably, it regularly hangs on bots etc. So doing it first just replaces one kind of instability with another.

**Comment by alexrp** [comment]
Even if that's the case, is getting a native backtrace not more important than getting a managed one in the context of debugging native crashes? By prioritizing GDB before `mono_walk_stack`, we have a better chance of that.

To me, this isn't so much about what's most reliable at producing *some* kind of output, but about what produces the most *useful* output in most cases.

**Comment by alexrp** [comment]
Btw, do you have any examples of those hangs? I'd be interested to see if there's something we could do about them (e.g. start a separate process that could kill GDB after a certain timeout).

**Comment by lewurm** [comment]
this PR is hard to review because multiple things are squashed into a single commit.

I like the stuff around `trying_*`, but I'm not sure about the ordering thing.  have you actually seen `mono_walk_stack` failing?

**Comment by vargaz** [comment]
I think we should keep the current ordering, and figure out why the stack walk fails, it shouldn't fail in most cases.

**Comment by alexrp** [comment]
> this PR is hard to review because multiple things are squashed into a single commit.

I think it would be either way due to all the moving around of code, especially due to all the `#ifdef`s in that function. I'd just take a look at the code rather than the diff.

> I like the stuff around `trying_*`, but I'm not sure about the ordering thing. have you actually seen `mono_walk_stack` failing?

* https://jenkins.mono-project.com/job/test-mono-profilerstresstests/302/testReport/junit/MonoTests/profiler-stress/ahcbench/
* https://jenkins.mono-project.com/job/test-mono-profilerstresstests/302/testReport/MonoTests/profiler-stress/gosharp_regexp/
* https://jenkins.mono-project.com/job/test-mono-profilerstresstests/300/testReport/junit/MonoTests/profiler-stress/fsharp/

Basically all occurrences of this assertion on the profiler stress lane.

> I think we should keep the current ordering, and figure out why the stack walk fails, it shouldn't fail in most cases.

Sure, maybe not in most cases, but the particular crash that prompted me to open this PR is almost certainly stack corruption (since there is no actual way for `unwind_options` to be non-zero when called from the profiler's sampling signal handler). If we're dealing with stack corruption, all bets are off, and I'd actually be more surprised if `mono_walk_stack` *did* work.

**Comment by alexrp** [comment]
It's worth pointing out that we have a script in place on the bots to kill GDB sessions that have been running for longer than 5 minutes.

**Comment by lewurm** [comment]
I know that this is the case for jenkins, but is this also true on wrench?

**Comment by alexrp** [comment]
No idea. Have we had problems with LLDB hanging on OS X?

**Comment by akoeplinger** [comment]
I wouldn't worry about Wrench, it has a "no output received in last X minutes -> terminate process" behavior and we should be moving away from it anyway.

**Comment by vargaz** [comment]
We had gdb/lldb hang a lot of times on wrench, its not a 100% stable, so I think we should keep the current order, the managed stack trace creation fails way less often.
